File: scraper.js
const { chromium } = require('playwright');
const fs = require('fs');

const EMAIL_REGEX = /[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}/g;

class LinkedInScraper {
  constructor() {
    this.browser = null;
    this.context = null;
    this.authStatePath = 'auth_state.json';
  }

  async init() {
    this.browser = await chromium.launch({
      headless: false,
      args: ['--disable-blink-features=AutomationControlled', '--no-sandbox']
    });

    if (fs.existsSync(this.authStatePath)) {
      this.context = await this.browser.newContext({
        storageState: this.authStatePath,
        userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36'
      });
    } else {
      this.context = await this.browser.newContext({
        userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36'
      });
    }
  }

  async login(email, password) {
    const page = await this.context.newPage();
    try {
      await page.goto('https://www.linkedin.com/login', { waitUntil: 'domcontentloaded', timeout: 30000 });
      await page.waitForTimeout(2000);

      if (page.url().includes('/login') || page.url().includes('/checkpoint')) {
        await page.fill('#username', email);
        await page.waitForTimeout(500);
        await page.fill('#password', password);
        await page.waitForTimeout(500);
        await page.click('button[type="submit"]');
        console.log('LinkedIn login submitted. Waiting for redirect or 2FA...');
        await page.waitForURL('**/feed/**', { timeout: 90000 }).catch(() => {});
        await this.context.storageState({ path: this.authStatePath });
        console.log('Session saved to auth_state.json');
      }
    } catch (err) {
      console.error('Login error:', err.message);
    } finally {
      await page.close();
    }
  }

  async scrapePostsFromUrl(url, logCallback, maxPages = 10) {
    const page = await this.context.newPage();
    const results = [];
    const seenUrns = new Set();

    try {
      logCallback(`Navigating → ${url}`);
      await page.goto(url, { waitUntil: 'domcontentloaded', timeout: 90000 });
      await page.waitForTimeout(3000);

      // Check if redirected to login
      if (page.url().includes('/login') || page.url().includes('/checkpoint')) {
        logCallback('Error: LinkedIn session expired. Please re-login.');
        return results;
      }

      let pageNum = 1;
      let staleCount = 0;
      const MAX_STALE = 4;

      logCallback(`Page ${pageNum} loaded.`);

      while (pageNum <= maxPages) {
        // Expand all "see more" on current view
        await this._expandSeeMore(page);
        await page.waitForTimeout(800);

        // Extract posts
        const posts = await this._extractPosts(page);
        let newCount = 0;

        for (const post of posts) {
          if (!post.urn || seenUrns.has(post.urn)) continue;
          seenUrns.add(post.urn);
          newCount++;

          const emails = this._extractEmails(post.text);
          if (emails.length > 0) {
            logCallback(`Found: ${post.author || 'Unknown'} → ${emails.join(', ')}`);
            results.push({
              content: post.text,
              author: post.author,
              authorTitle: post.authorTitle,
              emails
            });
          }
        }

        logCallback(`Page ${pageNum}: +${newCount} new | ${seenUrns.size} total seen | ${results.length} with emails`);

        // Try paginated Next button first
        const nextBtn = await page.$(
          'button[aria-label="Next"]:not([disabled]), .artdeco-pagination__button--next:not([disabled])'
        );

        if (nextBtn) {
          logCallback(`Clicking Next → page ${pageNum + 1}...`);
          await nextBtn.scrollIntoViewIfNeeded();
          await nextBtn.click();
          await page.waitForTimeout(3500);
          await page.waitForLoadState('domcontentloaded').catch(() => {});
          await page.waitForTimeout(1500);
          pageNum++;
          staleCount = 0;
          logCallback(`Page ${pageNum} loaded.`);
          continue;
        }

        // Infinite scroll mode
        const prevHeight = await page.evaluate(() => document.body.scrollHeight);

        // Scroll in steps to trigger lazy loads
        await page.evaluate(() => {
          window.scrollBy(0, Math.floor(window.innerHeight * 0.8));
        });
        await page.waitForTimeout(1200);
        await page.evaluate(() => window.scrollTo(0, document.body.scrollHeight));
        await page.waitForTimeout(1800);

        const newHeight = await page.evaluate(() => document.body.scrollHeight);

        if (newHeight === prevHeight && newCount === 0) {
          staleCount++;
          logCallback(`No new content (${staleCount}/${MAX_STALE})`);
          if (staleCount >= MAX_STALE) {
            logCallback('End of feed reached.');
            break;
          }
          await page.waitForTimeout(2000);
        } else {
          staleCount = 0;
          if (seenUrns.size % 10 === 0 && seenUrns.size > 0) {
            pageNum++;
            logCallback(`--- Page ${pageNum} (scroll) ---`);
          }
        }
      }

      logCallback(`Done. ${results.length} posts with emails across ${pageNum} page(s).`);
      return results;

    } catch (err) {
      logCallback(`Scrape error: ${err.message}`);
      return results;
    } finally {
      await page.close();
    }
  }

  async _expandSeeMore(page) {
    try {
      // Multiple selector patterns for different LinkedIn layouts
      const selectors = [
        'button.lt-line-clamp__more',
        'button[aria-label="see more"]',
        'span.lt-line-clamp__more',
        '.feed-shared-inline-show-more-text button',
        '.see-more'
      ];
      for (const sel of selectors) {
        const btns = await page.$$(sel);
        for (const btn of btns) {
          await btn.click().catch(() => {});
          await page.waitForTimeout(80);
        }
      }
    } catch (_) {}
  }

  async _extractPosts(page) {
    return page.evaluate(() => {
      const results = [];
      const seen = new Set();

      // Primary selector: data-urn posts
      const nodes = document.querySelectorAll('[data-urn]');
      for (const el of nodes) {
        const urn = el.getAttribute('data-urn');
        if (!urn || seen.has(urn)) continue;
        seen.add(urn);

        const authorEl = el.querySelector(
          '.update-components-actor__name span[aria-hidden="true"], ' +
          '.feed-shared-actor__name span[aria-hidden="true"], ' +
          '.update-components-actor__name, ' +
          '.feed-shared-actor__name'
        );
        const titleEl = el.querySelector(
          '.update-components-actor__description span[aria-hidden="true"], ' +
          '.feed-shared-actor__description span[aria-hidden="true"], ' +
          '.update-components-actor__description, ' +
          '.feed-shared-actor__description'
        );
        const textEl = el.querySelector(
          '.update-components-text, ' +
          '.feed-shared-update-v2__description, ' +
          '.feed-shared-text, ' +
          '.attributed-text-segment-list__container'
        );

        const text = textEl ? textEl.innerText.trim() : el.innerText.trim();
        if (!text) continue;

        results.push({
          urn,
          author: authorEl ? authorEl.innerText.trim().split('\n')[0] : '',
          authorTitle: titleEl ? titleEl.innerText.trim().split('\n')[0] : '',
          text
        });
      }

      return results;
    });
  }

  _extractEmails(text) {
    const emailRegex = /[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}/g;
    const matches = text.match(emailRegex) || [];
    // Filter out image/media filenames that match the regex
    const filtered = matches.filter(e => !e.match(/\.(png|jpg|jpeg|gif|svg|webp)$/i));
    return [...new Set(filtered)];
  }

  async close() {
    if (this.browser) await this.browser.close();
    this.browser = null;
    this.context = null;
  }
}

module.exports = LinkedInScraper;

File: scraper.py
import re
import os
import json
import asyncio
import random
from playwright.async_api import async_playwright

EMAIL_REGEX = re.compile(r'[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}')

class LinkedInScraper:
    def __init__(self):
        self.browser = None
        self.context = None
        self.auth_state_path = 'auth_state.json'
        self.playwright = None

    async def init(self, headless=False):
        self.playwright = await async_playwright().start()
        self.browser = await self.playwright.chromium.launch(
            headless=headless,
            args=['--disable-blink-features=AutomationControlled', '--no-sandbox']
        )
        
        if os.path.exists(self.auth_state_path):
            self.context = await self.browser.new_context(
                storage_state=self.auth_state_path,
                user_agent='Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36'
            )
        else:
            self.context = await self.browser.new_context(
                user_agent='Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36'
            )

    async def login(self, email, password):
        pass

    async def scrape_posts_from_url(self, url, log_callback, post_callback, max_pages=10):
        page = await self.context.new_page()
        results = []
        seen_urns = set()
        
        try:
            log_callback(f"Navigating -> {url}")
            await page.goto(url, wait_until='domcontentloaded', timeout=90000)
            await page.wait_for_timeout(3000)
            
            if '/login' in page.url or '/checkpoint' in page.url:
                log_callback('Authentication required. Please log in manually in the browser window...')
                try:
                    # Wait for user to log in and get redirected away from login/checkpoint pages
                    await page.wait_for_url(lambda u: '/login' not in u and '/checkpoint' not in u, timeout=300000)
                    log_callback('Authenticated! Saving session state...')
                    await self.context.storage_state(path=self.auth_state_path)
                    # Re-navigate to target URL now that we are authenticated
                    await page.goto(url, wait_until='domcontentloaded', timeout=90000)
                    await page.wait_for_timeout(3000)
                except Exception:
                    log_callback('Error: Login timeout or verification failed. Stopping.')
                    return results
                
            page_num = 1
            stale_count = 0
            max_stale = 4
            
            log_callback(f"Page {page_num} loaded.")
            
            while page_num <= max_pages:
                # Wait for search results or post cards to load dynamically via AJAX
                try:
                    await page.wait_for_selector('[role="listitem"], [data-urn], [componentkey*="FeedType"]', timeout=15000)
                    await page.wait_for_timeout(1000)
                except Exception:
                    pass

                # Expand "see more" buttons
                await self._expand_see_more(page)
                await page.wait_for_timeout(800)
                
                # Extract posts
                posts = await self._extract_posts(page)
                new_count = 0
                
                for post in posts:
                    urn = post.get('urn')
                    if not urn or urn in seen_urns:
                        continue
                    seen_urns.add(urn)
                    new_count += 1
                    
                    emails = self._extract_emails(post.get('text', ''))
                    if emails:
                        log_callback(f"Found URN {urn}: {post.get('author') or 'Unknown'} -> {', '.join(emails)}")
                    
                    try:
                        await post_callback(post, emails)
                    except Exception as ex:
                        log_callback(f"Callback error: {str(ex)}")
                
                # Try paginated Next button first
                next_btn = await page.query_selector(
                    'button[aria-label="Next"]:not([disabled]), .artdeco-pagination__button--next:not([disabled])'
                )
                
                if next_btn:
                    await page.wait_for_timeout(random.randint(1500, 3000))
                    log_callback(f"Clicking Next -> page {page_num + 1}...")
                    await next_btn.scroll_into_view_if_needed()
                    await next_btn.click()
                    await page.wait_for_timeout(random.randint(4000, 7000))
                    try:
                        await page.wait_for_load_state('domcontentloaded', timeout=5000)
                    except Exception:
                        pass
                    await page.wait_for_timeout(random.randint(2000, 4000))
                    page_num += 1
                    stale_count = 0
                    log_callback(f"Page {page_num} loaded.")
                    continue
                    
                # Infinite scroll mode
                prev_height = await page.evaluate("() => document.body.scrollHeight")
                
                await page.evaluate("() => window.scrollBy(0, Math.floor(window.innerHeight * 0.8))")
                await page.wait_for_timeout(random.randint(1500, 2500))
                await page.evaluate("() => window.scrollTo(0, document.body.scrollHeight)")
                await page.wait_for_timeout(random.randint(2000, 3500))
                
                new_height = await page.evaluate("() => document.body.scrollHeight")
                
                if new_height == prev_height and new_count == 0:
                    stale_count += 1
                    log_callback(f"No new content ({stale_count}/{max_stale})")
                    if stale_count >= max_stale:
                        log_callback('End of feed reached.')
                        break
                    await page.wait_for_timeout(2000)
                else:
                    stale_count = 0
                    if len(seen_urns) % 10 == 0 and len(seen_urns) > 0:
                        page_num += 1
                        log_callback(f"--- Page {page_num} (scroll) ---")
                        
            log_callback(f"Done. {len(results)} posts with emails across {page_num} page(s).")
            return results
        except Exception as e:
            log_callback(f"Scrape error: {str(e)}")
            return results
        finally:
            await page.close()

    async def _expand_see_more(self, page):
        try:
            selectors = [
                'button.lt-line-clamp__more',
                'button[aria-label="see more"]',
                'span.lt-line-clamp__more',
                '.feed-shared-inline-show-more-text button',
                '.see-more'
            ]
            for sel in selectors:
                btns = await page.query_selector_all(sel)
                for btn in btns:
                    try:
                        await btn.click()
                        await page.wait_for_timeout(random.randint(400, 1200))
                    except Exception:
                        pass
        except Exception:
            pass

    async def _extract_posts(self, page):
        return await page.evaluate("""() => {
            const results = [];
            const seen = new Set();
            const nodes = document.querySelectorAll('[role="listitem"], [data-urn], [componentkey*="FeedType"]');
            for (const el of nodes) {
                let urn = el.getAttribute('data-urn') || el.getAttribute('componentkey');
                if (!urn) {
                    const textContent = el.innerText || '';
                    if (textContent.length < 50) continue;
                    urn = textContent.slice(0, 100);
                }
                if (seen.has(urn)) continue;
                seen.add(urn);

                const authorLink = el.querySelector('a[href*="/in/"]');
                let author = '';
                if (authorLink) {
                    author = authorLink.innerText.trim().split('\\n')[0];
                }
                if (!author) {
                    const fallbackActor = el.querySelector('.update-components-actor__name, .feed-shared-actor__name');
                    if (fallbackActor) {
                        author = fallbackActor.innerText.trim().split('\\n')[0];
                    }
                }

                let authorTitle = '';
                const titleEl = el.querySelector(
                    '.update-components-actor__description, ' +
                    '.feed-shared-actor__description'
                );
                if (titleEl) {
                    authorTitle = titleEl.innerText.trim().split('\\n')[0];
                } else if (authorLink) {
                    const parentContainer = authorLink.parentElement;
                    if (parentContainer) {
                        const descriptionSibling = parentContainer.querySelector('p, div');
                        if (descriptionSibling && descriptionSibling !== authorLink) {
                            authorTitle = descriptionSibling.innerText.trim().split('\\n')[0];
                        }
                    }
                }

                const textEl = el.querySelector(
                    '[data-testid="expandable-text-box"], ' +
                    '.update-components-text, ' +
                    '.feed-shared-update-v2__description, ' +
                    '.feed-shared-text, ' +
                    '.attributed-text-segment-list__container'
                );

                const text = textEl ? textEl.innerText.trim() : el.innerText.trim();
                if (!text || text.length < 10) continue;

                results.push({
                    urn: urn,
                    author: author,
                    authorTitle: authorTitle,
                    text: text
                });
            }
            return results;
        }""")

    def _extract_emails(self, text):
        matches = EMAIL_REGEX.findall(text) or []
        filtered = [e for e in matches if not re.search(r'\.(png|jpg|jpeg|gif|svg|webp)$', e, re.IGNORECASE)]
        return list(dict.fromkeys(filtered))

    async def close(self):
        try:
            if self.browser:
                await self.browser.close()
        except Exception:
            pass
        self.browser = None
        self.context = None
        if self.playwright:
            try:
                await self.playwright.stop()
            except Exception:
                pass
        self.playwright = None
